### DISCUSSION TITLE
Nix flake - How do I use eww-wayland?

### DISCUSSION THREAD

**Discussion by Kodlak15** [elkowar/eww#961] [Q&A] [answered]
Hello there. I am brand new to Nix and NixOS and am trying to get eww set up using the flake. I have successfully installed eww through home-manager, but when I attempt to open my bar I get an error that states eww was not compiled with support for Wayland. 

Here is how I am importing eww in my flake.nix:
```
inputs = {
  ...
  ...
  eww = {
    url = "github:elkowar/eww";
  };
};
```

I include eww in the output parameters in my flake.nix:
```
outputs = {
  ...
  ...
  eww
};
```

This is the relevant snippet from my home configuration:
```
programs.eww = {
  enable = true;
  configDir = "/path/to/my/config";
};
```

I was taking a look at the code within the flake and noticed an override that seems like it ought to be able to handle this issue, but I have not a clue how to use it. If somebody could walk me through how I can deal with this I would really appreciate it!

**Comment by w-lfchen** [comment]
I don't really know what your flake is supposed to do (output your system configuration?), but I'd suggest that you use the eww-wayland package from [nixpkgs](https://search.nixos.org/packages?channel=23.05&show=eww-wayland&from=0&size=50&sort=relevance&type=packages&query=eww). You can do that by adding `package = <nixpkgs input name>.eww-wayland;` to your home manager configuration.
Incase your flake is for setting up your system, you can remove eww from it, as home-manager will install it for you already.

**Reply by Kodlak15** [reply]
Sorry, I could have been more clear. My flake does indeed output my system configuration, as well as my home configuration. 

I suppose the practical decision would be to just use the nixpkgs version, at least for the time being. I was trying to install eww following the same pattern I did to install Hyprland and NixVim. In both of those cases it is suggested to use each of their respective flakes as inputs in your flake.nix. I noticed this override within the eww flake that appears as though it is intended to resolve my problem:

```
eww-wayland = final.eww.override { withWayland = true; };
```

I would like to know how to use it as the knowledge ought to be applicable to other situations as well. Or perhaps I am misunderstanding things. This is a lot of new stuff to learn at once. Appreciate your comment and your help!

**Comment by Kodlak15** [comment, marked as answer]
I managed to figure this out. Referring to this individuals dotfiles was very helpful: https://github.com/fufexan/dotfiles. I will explain what ended up working for me for anyone else who may become confused by this. 

In your flake.nix add eww and the rust overlay as inputs. The rust overlay seems to be necessary so that the nightly toolchain is available to build eww. 
```
# Flake.nix
{
  inputs = {
    eww = {
      url  = "github:elkowar/eww";
      inputs.nixpkgs.follows = "nixpkgs";
      inputs.rust-overlay.follows = "rust-overlay";
    };

    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  ...
  ...
  ...
}
```

Enable eww in your home configuration like like so:
```
{inputs, pkgs, ...}:
{
  programs.eww = {
    enable = true;
    package = inputs.eww.packages.${pkgs.system}.eww-wayland;
    configDir = ./config; # path to your configuration
  };
}
```
